File: python/agent/langgraph-fsi-agent/before/src/galileo_langgraph_fsi_agent/tools/pinecone_retrieval_tool.py
"""
A tool for retrieving information from the Pinecone vector database.
"""

from typing_extensions import override

from langchain.tools import BaseTool
from langchain_pinecone import PineconeVectorStore
from langchain_openai import OpenAIEmbeddings
from pydantic import BaseModel, Field


class RetrievalInput(BaseModel):
    """
    RetrievalInput is a Pydantic model representing the input schema for a document retrieval operation.
    Attributes:
        query (str): The search query used to find relevant documents.
        k (int, optional): The number of documents to retrieve. Defaults to 3.
    """

    query: str = Field(description="The search query to find relevant documents")
    k: int = Field(default=3, description="Number of documents to retrieve")


class PineconeRetrievalTool(BaseTool):
    """
    PineconeRetrievalTool is a tool for retrieving relevant information from a financial services knowledge base using Pinecone as a vector store.
    """

    name: str = "pinecone_retrieval"
    description: str = "Retrieve relevant information from the financial services knowledge base"
    args_schema: type[BaseModel] = RetrievalInput  # type: ignore

    def __init__(self, index_name: str):
        super().__init__()
        self._embeddings = OpenAIEmbeddings()
        self._index_name = index_name
        self._vector_store = PineconeVectorStore(index_name=self._index_name, embedding=self._embeddings)

    @override
    def _run(self, query: str, k: int = 3) -> str:
        """Execute the retrieval"""
        try:
            # Perform similarity search
            results = self._vector_store.similarity_search(query, k=k)

            if not results:
                return "No relevant information found in the knowledge base."

            # Format the results
            formatted_results = []
            for i, doc in enumerate(results, 1):
                formatted_results.append(f"Document {i}:\n{doc.page_content}\n")

            return "\n".join(formatted_results)

        except Exception as e:
            return f"Error retrieving information: {str(e)}"

File: typescript/agent/langgraph-fsi-agent/tools/pineconeRetrievalTool.ts
/**
 * A tool for retrieving information from the Pinecone vector database.
 */
import { PineconeStore } from "@langchain/pinecone";
import { OpenAIEmbeddings } from "@langchain/openai";
import { z } from 'zod';

import { Pinecone } from "@pinecone-database/pinecone";
import { tool } from "@langchain/core/tools";

// Load environment variables from .env file so we have the Pinecone API key
import dotenv from 'dotenv';
dotenv.config();

/**
 * RetrievalInput is a Zod schema representing the input for a document retrieval operation.
 * Attributes:
 *   query (string): The search query used to find relevant documents.
 *   k (number, optional): The number of documents to retrieve. Defaults to 3.
 */
const RetrievalInputSchema = z.object({
    query: z.string().describe('The search query to find relevant documents'),
    k: z.number().default(3).describe('Number of documents to retrieve'),
});


/**
 * A retriever class that interfaces with a Pinecone vector store to perform semantic search
 * using OpenAI embeddings. It allows querying the vector store for documents similar to a given query.
 *
 * @remarks
 * This class initializes the OpenAI embeddings and connects to a Pinecone index via the PineconeStore.
 * The `run` method executes a similarity search and returns the most relevant documents' content.
 *
 * @example
 * ```typescript
 * const retriever = new PineconeRetriever('my-index');
 * const result = await retriever.run({ query: 'What is LangChain?', k: 3 });
 * console.log(result);
 * ```
 */
class PineconeRetriever {

    private embeddings: OpenAIEmbeddings;
    private vectorStore: PineconeStore;

    constructor(indexName: string) {
        this.embeddings = new OpenAIEmbeddings();
        const pinecone = new Pinecone();
        this.vectorStore = new PineconeStore(this.embeddings, { pineconeIndex: pinecone.Index(indexName) });
    }

    /**
     * Execute the retrieval.
     * @param input RetrievalInput
     * @returns Promise<string>
     */
    async run(input: any): Promise<string> {
        try {
            const { query, k } = input;
            const results = await this.vectorStore.similaritySearch(query, k);

            if (!results || results.length === 0) {
                return 'No relevant information found in the knowledge base.';
            }

            const formattedResults = results.map((doc, i) => `Document ${i + 1}:\n${doc.pageContent}\n`);
            return formattedResults.join('\n');
        } catch (e: any) {
            return `Error retrieving information: ${e.message || String(e)}`;
        }
    }
}

// Create an instance of the PineconeRetriever with the index name for credit card information
const pineconeRetrievalTool = new PineconeRetriever("credit-card-information");

/**
 * Define a tool for retrieving information from the Pinecone vector database.
 * It extends the ClientTool class.
 */
// @ts-ignore - Complex type instantiation issue with LangChain tool function
export const creditCardInformationRetrievalTool = tool(
    async (input): Promise<string> => {
        try {
            return await pineconeRetrievalTool.run(input);
        } catch (e: any) {
            return `Error retrieving information: ${e.message || String(e)}`;
        }
    },
    {
        name: 'credit_card_information_retrieval',
        description: 'Retrieve relevant information about credit card options from the Brahe Bank knowledge base.',
        schema: RetrievalInputSchema,
    }
);
